nrf_wifi/fw_if/umac_if/src/fmac_api_common.c

```diff
@@ -24,6 +24,177 @@
 
+#include <patch_info.h>
+
+static int nrf_wifi_patch_version_compat(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+				const unsigned int version)
+{
+	unsigned int family, major, minor, patch;
+
+	family = (version >> 24) & 0xff;
+	major = (version >> 16) & 0xff;
+	minor = (version >> 8) & 0xff;
+	patch = (version >> 0) & 0xff;
+
+	if (family != RPU_VERSION) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Incompatible RPU version: %d", family);
+		return -1;
+	}
+
+	if (major != RPU_MAJOR_VERSION) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Incompatible RPU major version: %d", major);
+		return -1;
+	}
+
+	/* TODO: Allow minor version to be different */
+	if (minor != RPU_MINOR_VERSION) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Incompatible RPU minor version: %d", minor);
+		return -1;
+	}
+
+	/* TODO: Allow patch version to be different */
+	if (patch != RPU_EXTRA_VERSION) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Incompatible RPU patch version: %d", patch);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int nrf_wifi_patch_feature_flags_compat(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+				const unsigned int feature_flags)
+{
+#ifdef CONFIG_NRF700X_RADIO_TEST
+	if (!(feature_flags & NRF70_FEAT_RADIO_TEST)) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Radio test feature flag not set");
+		return -1;
+	}
+#elif defined(CONFIG_NRF700X_SCAN_ONLY)
+	if (!(feature_flags & NRF70_FEAT_SCAN_ONLY)) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Scan only feature flag not set");
+		return -1;
+	}
+#elif defined(CONFIG_NRF700X_SYSTEM_MODE)
+	if (!(feature_flags & NRF70_FEAT_SYSTEM_MODE)) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"System mode feature flag not set");
+		return -1;
+	}
+#else
+	nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+		"Invalid feature flags: 0x%x or build configuration",
+		feature_flags);
+#endif
+
+	return 0;
+}
+
+enum nrf_wifi_status nrf_wifi_fmac_fw_parse(struct nrf_wifi_fmac_dev_ctx *fmac_dev_ctx,
+					   const void *fw_data,
+					   unsigned int fw_size,
+					   struct nrf_wifi_fmac_fw_info *fw_info)
+{
+	struct nrf70_fw_image_info  *info = (struct nrf70_fw_image_info *)fw_data;
+	int offset = 0;
+	int image_id;
+
+	if (!fw_data || !fw_size || !fw_info) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Invalid parameters");
+		return -1;
+	}
+
+	if (fw_size < sizeof(struct nrf70_fw_image_info)) {
+		nrf_wifi_osal_log_err(fmac_dev_ctx->fpriv->opriv,
+			"Invalid fw_size: %d, minimum size: %d",
+			fw_size, sizeof(struct nrf70_fw_image_info));
+		return -1;
+	}
+
+	nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+		"num_images: %d", info->num_images);
+	nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+		"version: 0x%x", info->version);
+	nrf_wifi_osal_log_dbg(fmac_dev_ctx->fpriv->opriv,
+		"feature_flags: %d", info->feature_flags);
+
+	if (info->num_images != NRF_WIFI_PATCH_NUM_IMAGES) {
```

Comment:
Is this check always true? As in can't we have only UMAC or LMAC patch?

Reply:
The `NRF_WIFI_PATCH_NUM_IMAGES` is defined in the `patch_info.h` and is controlled by RPU. We just do a sanity check against the binary.